File: app/core/agent/tools/compact.ts
/** compact tool — DEPRECATED stub (Phase 2 Task 2.21): placeholder keeping the SDK toolset complete during migration; compaction becomes orchestrator-driven (not model-invoked) and Phase 3 Task 3.6 deletes this + its registry entries entirely. */

import { tool } from 'ai';
import { z } from 'zod';
import type { ToolResult, ToolRegistration } from './types';
import type { ToolContext } from './tool-context';
import { withPermission } from '../permission-wrapper';

export async function runCompact(keepLast: number): Promise<ToolResult> {
  return {
    status: 'executed',
    output: 'Done. Continue with your current task.',
    meta: `keep last ${keepLast}`,
  };
}

export const compactTool: ToolRegistration = {
  name: 'compact',
  definition: {
    name: 'compact',
    description:
      '[Internal] Summarize earlier conversation history. The orchestrator ' +
      'handles this automatically — this tool exists for edge cases only.',
    input_schema: {
      type: 'object',
      properties: {
        keep_last: {
          type: 'number',
          description: 'Number of most-recent messages to keep verbatim. Older ones get summarized. Default 6.',
        },
      },
    },
  },
  riskTier: 'read_only',
  requiresWorktree: false,
  timeoutMs: 1_000,
  autoApproveIn: ['plan', 'ask', 'edit', 'full'],
  execute: async (args, _ctx) => runCompact(typeof args.keep_last === 'number' ? args.keep_last : 6),
};

// ─── SDK factory (deprecation stub — deleted in Phase 3 Task 3.6) ──────

export function createCompactTool(ctx: ToolContext) {
  return tool({
    description:
      '[Internal] Summarize earlier conversation history. The orchestrator ' +
      'handles this automatically — this tool exists for edge cases only.',
    inputSchema: z.object({
      keep_last: z.number().optional().describe('Number of most-recent messages to keep verbatim. Default 6.'),
    }),
    execute: async ({ keep_last }) =>
      withPermission(ctx, 'compact', { keep_last }, () =>
        runCompact(typeof keep_last === 'number' ? keep_last : 6),
      ),
  });
}

File: src-tauri/crates/tide-tools/src/tools/compact.rs
//! compact — port of `app/core/agent/tools/compact.ts` (). The TS
//! tool was a deprecation stub: compaction is orchestrator-driven, not
//! model-invoked, and the tool exists to keep the SDK toolset complete.
//! The orchestrator intercepts `compact` calls and runs the shared
//! auto-compact path (see the app crate's `agent::auto_compact`) before
//! returning this stub outcome; `Tool::execute` is the bare stub for
//! non-turn callers (drift tests, direct registry use).

use serde_json::json;

use crate::permission::RiskTier;
use crate::{Tool, ToolContext, ToolError, ToolOutcome, ToolSpec};

const DESCRIPTION: &str = "[Internal] Summarize earlier conversation history. The orchestrator handles this automatically \u{2014} this tool exists for edge cases only.";

pub const DEFAULT_KEEP_LAST: u64 = 6;

pub fn run_compact(keep_last: u64) -> ToolOutcome {
    ToolOutcome::executed("Done. Continue with your current task.")
        .with_meta(format!("keep last {keep_last}"))
}

pub struct CompactTool;

impl Tool for CompactTool {
    fn spec(&self) -> ToolSpec {
        ToolSpec {
            name: "compact".into(),
            description: DESCRIPTION.into(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "keep_last": {
                        "type": "number",
                        "description": "Number of most-recent messages to keep verbatim. Older ones get summarized. Default 6."
                    }
                }
            }),
        }
    }

    fn risk_tier(&self) -> RiskTier {
        RiskTier::ReadOnly
    }

    fn execute(
        &self,
        _ctx: &ToolContext,
        args: serde_json::Value,
    ) -> Result<ToolOutcome, ToolError> {
        let keep_last = crate::tools::arg_u64(&args, "keep_last").unwrap_or(DEFAULT_KEEP_LAST);
        Ok(run_compact(keep_last))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn stub_result_matches_the_ts_shape() {
        let out = run_compact(4);
        assert_eq!(out.status, crate::OutcomeStatus::Executed);
        assert_eq!(out.output, "Done. Continue with your current task.");
        assert_eq!(out.meta.as_deref(), Some("keep last 4"));

        let tmp = tempfile::tempdir().unwrap();
        let tool = CompactTool;
        assert_eq!(tool.spec().name, "compact");
        assert_eq!(tool.risk_tier(), RiskTier::ReadOnly);
        let out = tool
            .execute(&ToolContext::new(tmp.path()), json!({ "keep_last": 2 }))
            .unwrap();
        assert_eq!(out.meta.as_deref(), Some("keep last 2"));
        // Missing keep_last → default 6.
        let out = tool
            .execute(&ToolContext::new(tmp.path()), json!({}))
            .unwrap();
        assert_eq!(out.meta.as_deref(), Some("keep last 6"));
    }
}
